### PR TITLE
Docker Workflow

### LINKED ISSUE #4480: Best Practices with Containers (Docker)
Branched from #4062. This is a collection of best-practices of partitioning containers with spack until they are mature enough for a PR.

### Docker Specific

- `ENV DEBIAN_FRONTEND noninteractive`
- `ENV FORCE_UNSAFE_CONFIGURE 1`
- adding a non-privileged (system) user for installs since some installs check for root and abort, add sudoers rights for the app

### Dependencies

- start from a fixed version of a docker image (tag)
- start from a fixed version of spack (tag or sha)
- install build tools (which are not linked) directly via package manager: cmake, editors, pkg-config, modules/lnmod/dotkit, python
- GPU/CUDA: e.g. from `nvidia-cuda` images (with "spack-external" cuda package)

### Minimal layers

- concat all apt-get install targets + `&& rm -rf /var/lib/apt/lists/*`
- concat all spack install calls + `&& spack clean -a`
- `git clone --depth 50` or `git archive` for a *fixed* commit of `spack` (ideally: `spack bootstrap` #4062)

### Environment

- how to add `packages.yaml modules.yaml`, specifically for external packages
- setup of spack (path, module load, ...) env, e.g. lmod in a user profile file
- add user profile, e.g. to `/etc/profile.d/`

### Example with CUDA

PIConGPU 0.3.0 docker image with spack: https://github.com/ComputationalRadiationPhysics/picongpu/tree/dev/share/picongpu/dockerfiles

More examples: https://github.com/ax3l/dockerfiles

### Related issues

- Bootstrap to `/` #4062
- run spack as root user #5139

### PR DESCRIPTION
**Pull request by ax3l**
Close #4480: This adds a workflow section on how to use spack on Docker.

It provides an example on the best-practices I collected over the last months and circumvents the common pitfalls I tapped in.

Works with MPI, CUDA, Modules, execution as root, etc.

If you want to give it a swing, I pushed [an image](https://github.com/ax3l/dockerfiles/tree/master/spack/base) (without CUDA but MPI capable) of the latest `dev` to dockerhub. Just run
```bash
docker run -it ax3l/spack
```

and do `spack ...` magic!

Note: For the pushed image above, I already added a bit of convenience (+380MB) such as an editor, openssh server for MPI, installed `tar` via spack, etc. If you want to set up a CI job I would recommend to keep it more minimal, see the example for the docs in this PR.

Setting up a regularly pushed, minimal `spack:develop` image would be a great thing, since it helps a lot to check and fix package build errors.

Happy shipping! 🚢 

Background: developed initially for [PIConGPU](https://github.com/ComputationalRadiationPhysics/picongpu/tree/dev/share/picongpu/dockerfiles). Here are [more spack based dockerfiles](https://github.com/ax3l/dockerfiles/tree/master/spack) of mine.

P.S. Works also splendit on Singularity with docker image import.
```
Bootstrap: docker
From: ax3l/spack:latest

%runscript
exec /bin/bash -l
```

### CODE REVIEW

**Comment by junghans** [comment]
@tgamblin: I would be nice if we could auto-deploy the docker container to as the 4th step in travis.

**Comment by sknigh** [comment]
@tgamblin What about a minimalistic container? It would go a long way in rooting out missing dependencies. I was able to use this to make an interactive spack environment a few months ago.

``` Dockerfile
FROM python:2.7.13-alpine3.6

RUN apk update && \
    apk --no-cache add \
        argp-standalone \
        bash \
        gcc \
        gfortran \
        g++ \
        patch \
        curl \
        git \
        make \
        vim

RUN git clone http://github.com/LLNL/spack.git

...
```

**Comment by ax3l** [comment]
@sknigh actually a minimal Dockerfile needs a bit more than your snippet to work at least as root and with modules. Additionally, MPI functionality is considered core functionality for Spack. Please see the details with explanation in my PR for a minimal Dockerfile :)

(The only thing not minimal was the image I pushed for demonstration.)

**Comment by jcftang** [comment]
A centos based example would also be nice, I'd be happy to submit a PR for that if you don't to add a centos example.

**Comment by pramodskumbhar** [comment]
@ax3l : this will be useful, thank you!

I wonder if it is worth adding note about spack and `multiarch`. When I created [docker image](https://github.com/pramodskumbhar/perfengineering/blob/master/containers/ubuntu_16_04/Dockerfile)  first time with Spack, I wanted to reduce "build" time. "Early" users will try to use system installed packages like mpi, hdf5, python etc. But this doesn't work in all cases due to `multiarch` (packages get installed into `/usr/lib/arch-triplet-here`. And specifying external package in `packages.yaml` as `/usr` fails often.

This is *not* related to Docker but external packages in general. But I am mentioning it here as people will try it with Docker workflow as well.

**Comment by ax3l** [comment]
Thank you! 
That is a good point. If you check the PR's last sections, I already show how to use an external cuda to save space and build time. 

I also build with external cuda, pkg-config and python regularly in docker.

**Comment by pramodskumbhar** [comment]
> I also build with external cuda, pkg-config and python regularly in docker.

yes, I have used this as well. This doesn't fail for all builds or all packages but consider package which does something like : 

```
                options.append('-mpiinc=%s' % spec['mpi'].prefix.include)
                options.append('-mpilib=%s' % spec['mpi'].prefix.lib)
```

but now we don't have libraries/includes their.

Few days ago I see this issue with python (2 or 3) on ubuntu 16 where I was trying to use `spec['python'].libs` but then `libs()` in python package couldn't find libraries.

**Comment by ax3l** [comment]
> A centos based example would also be nice, I'd be happy to submit a PR for that if you don't to add a centos example.

@jcftang Agreed! If we add docker images being build in the CI in a follow up, we could e.g. do a ubuntu and a centos one. :)

**Comment by ax3l** [comment]
@pramodskumbhar the multiarch issue seems like a general, external package usage note that we should definitely add to the corresponding section (not Docker as you said). Do you want to provide a PR about your experience? Luckily, I did not face those issues yet :)

**Comment by pramodskumbhar** [comment]
Sure! I will open separate PR, will try to do this on weekend.

File: lib/spack/docs/workflows.rst [hunk 1]
```diff
@@ -1106,6 +1106,150 @@ The main points that are implemented below:
      - make -j 2
      - make test
 
+-----------------------------------
+Using Spack to Create Docker Images
+-----------------------------------
+
+Spack can be the ideal tool to set up images for Docker (and Singularity)
+images.
```
Comment (alalazo) [inline lib/spack/docs/workflows.rst]: typo: **images** for Docker (and Singularity) **images**.

File: lib/spack/docs/workflows.rst [hunk 2]
```diff
@@ -1106,6 +1106,150 @@ The main points that are implemented below:
...
+   Package installs are followed by a clean-up of the system package index,
+   to avoid outdated information and it saves space.
+
+#. Install spack in ``/usr/local``.
+   Add ``setup-env.sh`` to profile scripts, so commands in *login* shells
+   can use the whole spack functionality, including modules.
+
+#. Install an example package (``tar``).
+   As with system package managers above, ``spack install`` commands should be
+   concated with a ``&& spack clean -a`` in order to keep image sizes small.
```
Comment (alalazo) [inline lib/spack/docs/workflows.rst]: typo: concatenated

**Comment by hartzell** [comment]
I agree w/ @alalazo -- super useful!

**Pull request merged**